dxr/mime.py

```diff
@@ -14,11 +14,26 @@ def icon(path, is_binary=False):
     return class_name
 
 
-def decode_data(data, encoding_guess):
-    """Given str data, return an (is_text, data) tuple, where data is returned
-    as unicode if we think it's text and were able to determine an encoding for
-    it."""
-    if not is_binary_string(data[:1024]):
+def decode_block(mmap_block, encoding_guess):
+    """Given mmap'd block, return an (is_text, data) tuple, where data is
+    returned as unicode if we think the block is text and were able to
+    determine an encoding for it, otherwise a string."""
+    initial_portion = mmap_block[:4096]
+    if not is_binary_string(initial_portion):
+        # Move the cursor back to the start of the block.
+        mmap_block.seek(0)
+        return decode_data(mmap_block.read(mmap_block.size()),
+                           encoding_guess, False)
+    return False, initial_portion
```

Comment:
It makes me uncomfortable because there's no way for the caller to tell if the result has been truncated.